### ISSUE TITLE
`airetry!` replacing schema and original prompt if `AIGenerate` wasn't `run!` beforehand

### ISSUE THREAD

**Issue by pabvald** [svilupp/PromptingTools.jl#216]
hej, 

I was trying a paraphrasing prompt to obtain variations of a message: 
```julia 
using PromptingTools
using PromptingTools.Experimental.AgentTools
const PT = PromptingTools

schema = PT.AzureOpenAISchema() 
model = "gpt-35-turbo" 
tpl = PT.create_template(
      system="You are an expert in the English language. Given a sentence, provide a paraphrase suing synonyms or modifying the order of the words", 
      user="# Sentence:\n\n{{sentence}}"
)
sentence = "How are you doing, mate?" 
```
In order to use `airetry!` I have created a lazy `AIGenerate` type:

```julia 
aicall1 = AIGenerate(
    schema, 
    tpl; 
    sentence=sentence, 
    model=model,
    config=RetryConfig(; max_retries=2)
)
aicall1.conversation 
# 2-element Vector{PromptingTools.AbstractChatMessage}:
#   PromptingTools.SystemMessage("You are an expert in the English language. Given a sentence, provide a paraphrase suing synonyms or modifying the order of the words")
#   PromptingTools.UserMessage{String}("# Sentence:\n\n{{sentence}}", [:sentence], :usermessage)
```
When  running `aicall` once and running `airetry!` afterwards, it works as expected:
```julia
feedback = "The sentences are still the same, modify something" 

aicall1 |> run!
# [ Info: Tokens: 55 in 0.5 seconds
# AICall{typeof(aigenerate)}(Messages: 3, Success: true)
#  - Preview of the Latest AIMessage (see property `:conversation`):
#   Hello, friend, how are you?

airetry!(aicall1, feedback) do conv
    new_sentence = try 
        last_output(conv) 
    catch e 
        e 
    end 

    new_sentence == sentence # the correct condtion would be `new_sentence != sentence` but I want to force the retry
end 
# [ Info: Condition not met. Retrying...
# [ Info: Tokens: 86 in 0.3 seconds
# [ Info: Condition not met. Retrying...
# [ Info: Tokens: 128 in 0.3 seconds
# [ Info: Condition not met, but maximum retry budget was reached (Retries: 2/2, Calls: 3/99).
# AICall{typeof(aigenerate)}(Messages: 7, Success: false)
# - Preview of the Latest AIMessage (see property `:conversation`):
#  Hey there, buddy, how are you?


aicall1.conversation 
# 7-element Vector{PromptingTools.AbstractMessage}:
#  PromptingTools.SystemMessage("You are an expert in the English language. Given a sentence, provide a paraphrase suing synonyms or modifying the order of the words")
#  PromptingTools.UserMessage{String}("# Sentence:\n\nHow are you doing, mate?", [:sentence], :usermessage)
#  AIMessage("Hey, friend, how are you feeling?")
#  PromptingTools.UserMessage{String}("### Feedback from Evaluator\n\nThe sentence is the same, modify something.\n", [:feedback], :usermessage)
#  AIMessage("Mate, how do you fare?")
#  PromptingTools.UserMessage{String}("### Feedback from Evaluator\n\nThe sentence is the same, modify something.\n----------\n\nThe sentence is the same, modify something.\n", [:feedback], :usermessage)
#  AIMessage("Hey there, buddy, how are you?")
```
However, if I run `airetry!` directly, without running `run!` first, the provided prompt is ignored and the default one is used: 
```julia 
aicall2 = AIGenerate(
    schema, 
    tpl; 
    sentence=sentence, 
    model=model,
    config=RetryConfig(; max_retries=2)
)
feedback = "The sentences are still the same, modify something" 

airetry!(aicall2, feedback) do conv
    new_sentence = try 
        last_output(conv) 
    catch e 
        e 
    end 

    new_sentence == sentence # again, forcing the retry
end
# [ Info: Condition not met. Retrying...
# [ Info: Tokens: 23 in 0.4 seconds
# [ Info: Condition not met. Retrying...
# [ Info: Tokens: 58 in 0.3 seconds
# [ Info: Condition not met, but maximum retry budget was reached (Retries: 2/2, Calls: 2/99).
# AICall{typeof(aigenerate)}(Messages: 4, Success: false)
# - Preview of the Latest AIMessage (see property `:conversation`):
#  Absolutely! How may I be of assistance to you today?

aicall2.conversation
# 4-element Vector{PromptingTools.AbstractMessage}:
#  PromptingTools.SystemMessage("Act as a helpful AI assistant")
#  AIMessage("Of course! how can I assist you today?")
#  PromptingTools.UserMessage{String}("### Feedback from Evaluator\n\nThe sentence is the same, modify something.\n", [:feedback], :usermessage)
#  AIMessage("Absolutely! How may I be of assistance to you today?")

**Comment by svilupp** [comment]
Thanks for flagging. 

What you're describing is expected, but I think I should add more validations to make sure people don't use it "wrongly". 

But I do see another bug in the printed outputs (Not sure why the feedback is pasted twice in the same message.. I suspect something with `feedback_inplace` argument).

To your example:
- AIGenerate is a lazy type -> it does nothing until you call `run!`
- airetry! -> it doesn't call the model UNLESS it needs to do fixing. It first checks the condition.

So in your example `aicall2`, you have effectively not called the LLM, you're using so it's retrying from nothing...
- The MCTS (and airetry!) works as a tree, so it always looks for a best place to restart from.
- Because we haven't started (no `run!`), `aicall2.samples` is kinda `empty` (no messages) and hence the best node to retry from has no messages (this line: https://github.com/svilupp/PromptingTools.jl/blob/188ede9998ce9cc7bab1152b9af4c6c38239a5f9/src/Experimental/AgentTools/retry.jl#L298)
- so then when you re-run the model, you start with an empty message, which gets auto-populated with that canned System Message

In short, it's all related to running evaluation on AICall that's empty (nothing to evaluate). The right behavior for me to add is to give user an error that they need to call `run!` first.

Does it make sense? Or did I misunderstand you?

**Comment by pabvald** [comment]
Yes, that makes sense. Thank you for the explanation! 

Pretty amazing work you did there, I hadn't seen the combination of search with LLMs until now, but if definitely makes sense! You can close the issue from my side.

**Comment by svilupp** [comment]
I'll leave it open until I implement the validations to prevent the same issue for others!

On a separate note, this functionality was too early when I built it -- open source models were either too large/too slow or collapsing after 2-3 turns of conversations. Now it would be really powerful with the current models.

PS: Pro tip -> for best MCTS performance you NEED to have multiple samples for each thing (to pick the best one to retry from), so you should utilize `n_samples` argument. OpenAI (Azure) can generate all those extra samples at once (no delay), I'd recommend ~2-3 to start.

**Comment by svilupp** [comment]
The validation has been added, so this shouldn't happen again.

> But I do see another bug in the printed outputs (Not sure why the feedback is pasted twice in the same message.. I suspect something with feedback_inplace argument).

It wasn't a bug - it was a feature. I forgot that that's how I implemented and it looked "weird" only because we kept providing "constant" feedback.
I've adjusted the logic that this whole history of feedback for all ancestor MCTS samples will be added ONLY IF we're running in "in-place" mode, where feedback is added to the previous UserMessage (to prevent chat history from growing)
